Support stargz snapshotter for dev stages

ktock opened the pull request:
Related issues: 
- https://github.com/moby/buildkit/issues/1396
- https://github.com/containerd/stargz-snapshotter/pull/53

Throughout building steps, "pulling base image" is one of the time-consuming steps.

Recently, containerd community started "Stargz Snapshotter" (containerd/stargz-snapshotter) non-core subproject for speeding up image pulling. With this snapshotter, we can directly mount stargz-formatted (but still docker-compatible) image to nodes, which takes much shorter than downloading the whole of the image to nodes.

This commit solves the performance problem on pulling base images for "dev" stages using stargz snapshotter.

We introduced new RecordType "snapshot" for unexported base images. These images aren't guaranteed to have their whole contents in the content store but snapshots (possibly "remote" snapshots) are provided. Target (exported) base images don't have this type because its contents need to be fully available on the node for exportation.

We can test this patch with stargz snapshotter on [this branch](https://github.com/ktock/stargz-snapshotter/tree/buildkit) (Please also see [this instruction](https://github.com/ktock/stargz-snapshotter/blob/a4568426aff225bdc07718c53dc1a7dbf79e8f9e/script/buildkit/testing.md)).
We can see the benchmarking result with [sample image](https://github.com/ktock/stargz-snapshotter/blob/a4568426aff225bdc07718c53dc1a7dbf79e8f9e/script/buildkit/sample_sgz/Dockerfile) on https://github.com/containerd/stargz-snapshotter/pull/53#issuecomment-596995434.

This PR also intends to be a discussion thread for the design and implementation strategies of this feature so if there is a better one please tell us.

AkihiroSuda commented:
Can we have integration tests?

AkihiroSuda commented on cmd/buildkitd/main_oci_worker.go:
```diff
@@ -262,7 +275,29 @@ func ociWorkerInitializer(c *cli.Context, common workerInitializerOpt) ([]worker
+		newFunc := func(root string) (ctdsnapshot.Snapshotter, error) {
+			gopts := []grpc.DialOption{
+				grpc.WithInsecure(),
+				grpc.WithBackoffMaxDelay(3 * time.Second),
```
WithBackoffMaxDelay is deprecated

AkihiroSuda replied:
See https://github.com/containerd/containerd/pull/4053/files#diff-1b43e06d76b779c2d0cb9f11f9bf69f1L100

ktock replied:
Thanks for your feedback. I fixed it.

AkihiroSuda commented on cmd/buildkitd/main_oci_worker.go:
```diff
@@ -49,9 +55,13 @@ func init() {
 		},
+		cli.StringFlag{
+			Name:  "oci-worker-proxy-snapshotter-address",
+			Usage: "use proxy snapshotter at specified address",
```
With `unix://` or without?

ktock replied:
Thanks for your feedback. Explicitly described not to include `unix://` prefix and added small validation against the specified path.

tonistiigi replied:
imho `*-address` should have protocol, `*-path` or `*-socket` doesn't need to.

ktock commented:
I'm currently working on the benchmarking and let me share the [latest benchmarking results](https://github.com/containerd/stargz-snapshotter/actions/runs/54973588) for building [sample images](https://github.com/ktock/stargz-snapshotter/tree/bc036743440e6deb76cd67627b46b354ed92a1b6/script/buildkit/sample)(benchmarking scripts are [here](https://github.com/ktock/stargz-snapshotter/tree/buildkit/script/buildkit)). I'll work on integration tests soon.

- `legacy`: unpatched buildkit(f1ecc7824ebcd147a9bd9426d19999a7eb8e9bab) + taball base image
- `stargz`: this PR's buildkit + stargz-formatted base image
- `estargz`: this PR's buildkit + extended stargz-formatted base image

### With OCI worker

![result_oci](https://user-images.githubusercontent.com/43872416/76617753-4f530100-656a-11ea-9623-009a1eea4c75.png)

### With containerd worker

![result_containerd](https://user-images.githubusercontent.com/43872416/76617784-5e39b380-656a-11ea-96a1-6bd3aa9eff14.png)

ktock commented:
Added unit and integration testes and fixed some bugs to make them happy. Containerd hasn't released remote-snapshots-related patches yet(but already merged)  so current integration testes are for oci worker. After these patches are released, we can apply tests for containerd worker with another PR. PTAL:pray:

AkihiroSuda commented:
@tonistiigi WDYT?

ktock commented:
@tonistiigi Thanks for you feedbacks and pointer! #870 seems to have same motivation with this patch in terms of defering pulling layers until when it's actually needed.

From #870, 
> All the ops using inputs should expect that a input could be a remote and not a cache.ImmutableRef. If they receive a remote but need to access the data directly (for example exec op) they can call a method on it that would extract it and change the type to immutableref. This method would be where the pull now actually happens...

Now containerd doesn't require layers being locally available even to exec something on them if these layers are remote snapshots. So in this case, we can hopefully defer the pulling (= changing the type from remote to immutableref) even until export occurs and we don't need to pull nor have immutablerefs for all unexported layers, without inconsistency.

This patch initially aimed to enable bulidkit to use the containerd's remote snapshots from scratch but now I think this feature can be applied more simply on #870 as an advanced optimization as mentioned above. Do you have plans to move #870 forward?

tonistiigi commented:
> Now containerd doesn't require layers being locally available even to exec something on them if these layers are remote snapshots. So in this case, we can hopefully defer the pulling (= changing the type from remote to immutableref) even until export occurs and we don't need to pull nor have immutablerefs for all unexported layers, without inconsistency.

Yes, it is fine to run a container without all the snapshots/blobs locally if the implementation supports it. But at any point the record used by that container or any other part of the build should remain exportable. This also complicates caching a lot. It is possible that the cache needs to store the registry ref where the rest of the data can be pulled and loading from a cache needs to check if the new build still has access to the remote resource. 

> we don't need to pull nor have immutablerefs for all unexported layers, 

More likely I think the design is that immutableref can have this separate implementation. If you would make some parts of build work on something else than immutableref this would be a lot of exceptions.

> Do you have plans to move #870 forward?

I probably don't have time to work on it on near future. But I can give you design direction if you are interested. @hinshun and I have also been discussing something called `MergeOp` that is very much related and needs a proper proposal/design. The hardest part is that I don't want to break Moby implementation with this and that model that doesn't have content blobs makes the whole thing even more complicated.

tonistiigi commented:
@ktock #1475 is getting into a ready state. Feel free to take a look if it has everything you need and port this over to the logic where refs can request a data source if they only have partial data when they are loaded from cache.

ktock commented:
Thanks! I'll take a look at.

ktock commented:
I've ported this PR over the lazy ref logic (#1475). Thanks for the great work! Based on that logic, the new patch doesn't require changes on the frontend and remote snapshots seem to be able to be used with immutableRefs with keeping them safely exportable. The following is a summary of the design in the new patch.

- `immutableRef.Extract()` and `Mount()` don't unlazy the ref if it's snapshot can be provided as a remote snapshot from the underlying snapshotter.
- mutableRefs possibly have some lazy parents (immutableRefs) because `Extract()` method don't guarantee that the contents are unlazied.
- `cache.DescHandler` has a new field `SnapshotLabels` which contains [key-value pairs used by the underlying snapshotter for searching for remote snapshots](https://github.com/containerd/containerd/blob/master/docs/remote-snapshotter.md#passing-snapshotter-specific-information) corresponding to the blobs. By default, `*containerimage.puller` adds labels needed to query stargz blobs to the registry.

@tonistiigi @AkihiroSuda What do you think about the above design? Please tell me if I'm missing something.
I'll work on tests once we reach an agreement on the design.

tonistiigi commented:
cc @sipsma

tonistiigi commented:
@ktock panic in CI

```
sandbox.go:218: panic: runtime error: invalid memory address or nil pointer dereference
        sandbox.go:218: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xbc2820]
        sandbox.go:218: 
        sandbox.go:218: goroutine 595 [running]:
        sandbox.go:218: github.com/moby/buildkit/cache.(*immutableRef).prepareRemoteSnapshots(0xc0004e1580, 0x1351d60, 0xc0004c6d50, 0x0, 0xc0003cf320, 0x1, 0x1)
        sandbox.go:218: 	/src/cache/refs.go:420 +0x160
        sandbox.go:218: github.com/moby/buildkit/cache.(*immutableRef).Extract(0xc0004e1580, 0x1351d60, 0xc0004c6d50, 0x0, 0x0)
        sandbox.go:218: 	/src/cache/refs.go:394 +0x19c```

ktock commented:
@tonistiigi Thanks for the comment. Fixed the panic.

AkihiroSuda commented on source/containerimage/pull.go:
```diff
@@ -203,14 +205,31 @@ func (p *puller) CacheKey(ctx context.Context, g session.Group, index int) (cach
+			for _, desc := range p.manifest.Remote.Descriptors {
+				layers += fmt.Sprintf("%s,", desc.Digest.String())
+			}
+			if len(layers) >= 1 {
```
use `strings.TrimSuffix` to trim the extra `,`.

Or consider using `strings.Join`

ktock commented on cache/refs.go:
```diff
@@ -391,15 +391,66 @@ func (sr *immutableRef) Extract(ctx context.Context) (rerr error) {
 	return sr.extract(ctx, sr.descHandlers)
 }
 
+func (sr *immutableRef) prepareRemoteSnapshots(ctx context.Context, dhs DescHandlers) (bool, error) {
```
Thanks for the feedback! Fixed to use `fileghtcontrol.Group`.

ktock commented:
Thanks for the feedback. Added test for this PR.

ktock commented on client/client_test.go:
```diff
@@ -2049,6 +2051,107 @@ func testBuildPushAndValidate(t *testing.T, sb integration.Sandbox) {
+
+	// Check if image layers are lazy
+	var layers []ocispec.Descriptor
+	for _, layer := range manifest.Layers[:len(manifest.Layers)-1] {
```
Thanks for the comment. This is because the topmost layer (created by running a command on that stargz image) isn't lazy. I added comments about this and also added a check to make sure that the topmost layer isn't lazy.

sipsma replied:
Ohh duh, forgot that the topmost layer was created by run, makes sense, thanks
